common/grid_map_utils/src/polygon_iterator.cpp

```diff
@@ -0,0 +1,180 @@
+      y_intersections.begin(), y_intersections.end(),
+      origin.y() - (grid_map.getSize()(1) - 1) * grid_map.getResolution(), std::greater());
+    while (iter != y_intersections.end() && std::next(iter) != y_intersections.end()) {
+      iter = y_intersections.erase(iter);
+      iter = y_intersections.erase(iter);
+    }
+    y_intersections_per_line.push_back(y_intersections);
+  }
+  return y_intersections_per_line;
+}
+
+std::pair<int, int> PolygonIterator::calculateRowRange(
+  const std::vector<Edge> & edges, const grid_map::Position & origin,
+  const grid_map::GridMap & grid_map)
+{
+  const auto min_vertex_x = edges.back().second.x();
+  const auto max_vertex_x = edges.front().first.x();
+
+  const auto dist_min_to_origin = origin.x() - min_vertex_x + grid_map.getResolution();
+  const auto dist_max_to_origin = origin.x() - max_vertex_x + grid_map.getResolution();
+  const auto min_row = std::clamp(
+    static_cast<int>(dist_max_to_origin / grid_map.getResolution()), 0, grid_map.getSize()(0) - 1);
+  const auto max_row = std::clamp(
+    static_cast<int>(dist_min_to_origin / grid_map.getResolution()), 0, grid_map.getSize()(0) - 1);
+
+  return {min_row, max_row};
+}
+
+PolygonIterator::PolygonIterator(
+  const grid_map::GridMap & grid_map, const grid_map::Polygon & polygon)
+{
+  auto poly = polygon;
+  if (poly.nVertices() < 3) return;
+  // repeat the first vertex to get the last edge [last vertex, first vertex]
+  if (poly.getVertex(0) != poly.getVertex(poly.nVertices() - 1)) poly.addVertex(poly.getVertex(0));
+
+  const auto resolution = grid_map.getResolution();
+  const auto & map_size = grid_map.getSize();
+  const auto & start_idx = grid_map.getStartIndex();
+  const auto origin = [&]() {
+    grid_map::Position origin;
+    grid_map.getPosition(start_idx, origin);
+    return origin;
+  }();
+  // We make line scan left -> right / up -> down *in the index frame* (idx[0,0] is pos[up, left]).
+  // In the position frame, this corresponds to high -> low Y values and high -> low X values.
+  const std::vector<Edge> edges = calculateSortedEdges(poly);
+  if (edges.empty()) return;
+  const auto from_to_row = calculateRowRange(edges, origin, grid_map);
+  const auto y_intersections_per_line =
+    calculateIntersectionsPerLine(edges, from_to_row, origin, grid_map);
+
+  // calculate map indexes between pairs of intersections Y values on each X line
+  polygon_indexes_.reserve(
+    y_intersections_per_line.size() * static_cast<size_t>(grid_map.getSize()(1)));
+  const auto from_row = start_idx(0) + from_to_row.first;
+  for (size_t i = 0; i < y_intersections_per_line.size(); ++i) {
+    const auto & y_intersections = y_intersections_per_line[i];
+    auto row = static_cast<int>(from_row + i);
+    grid_map::wrapIndexToRange(row, map_size(0));
+    for (auto y_iter = y_intersections.cbegin(); y_iter != y_intersections.cend(); ++y_iter) {
+      const auto dist_from_origin = origin.y() - *y_iter + resolution;
```

Comment:
This line is equal to adding 1 index to from value.
```suggestion
      const auto dist_from_origin = origin.y() - *y_iter;
      const auto from_col =
        std::clamp(static_cast<int>(1.0 + dist_from_origin / resolution), 0, map_size(1) - 1);

```

Reply:
Thank you. I applied the change [513b347](https://github.com/autowarefoundation/autoware.universe/pull/943/commits/513b347d009c5bb3ac815ace1f17db44cfc8a5a5)